academic_observatory_workflows/workflows/pubmed_telescope.py

```diff
@@ -0,0 +1,1623 @@
+            reset_ftp_counter=self.reset_ftp_counter,
+            max_download_retry=self.max_download_retry,
+        )
+
+        set_task_state(success, kwargs["ti"].task_id, release=release)
+
+    def upload_downloaded_updatefiles(self, release: PubMedRelease, **kwargs):
+        """Upload downloaded updatefiles files to GCS."""
+
+        # Grab list of files to upload.
+        datafiles_to_upload = [
+            datafile.download_file_path for datafile in release.datafile_list if not datafile.baseline
+        ]
+
+        success = gcs_upload_files(
+            bucket_name=self.cloud_workspace.download_bucket,
+            file_paths=datafiles_to_upload,
+        )
+
+        set_task_state(success, kwargs["ti"].task_id, release=release)
+
+    def transform_updatefiles(self, release: PubMedRelease, **kwargs):
+        """
+        Transform the *.xml.gz files downloaded from PubMed's FTP server into usable json-like files for BigQuery import.
+
+        This is a multithreaded task that pulls the information from the files using the keys defined in self.entity_list.
+        """
+
+        # Get list of datafiles for the baseline to trasform.
+        datafiles_to_transform = [datafile for datafile in release.datafile_list if not datafile.baseline]
+
+        # Process files in batches so that ProcessPoolExecutor doesn't deplete the system of memory
+        for i, chunk in enumerate(get_chunks(input_list=datafiles_to_transform, chunk_size=self.max_processes)):
+            with ProcessPoolExecutor(max_workers=self.max_processes) as executor:
+                logging.info(f"In chunk {i} and processing files: {chunk}")
+
+                futures = [
+                    executor.submit(
+                        transform_pubmed_xml_file_to_jsonl,
+                        datafile,
+                        [self.entity_list["upsert"], self.entity_list["delete"]],
+                    )
+                    for datafile in chunk
+                ]
+
+                # Make sure that all datafiles have been properly transformed.
+                for future in as_completed(futures):
+                    transformed_datafile = future.result()
+
+                    logging.info(f"Successfully transformed file - {transformed_datafile.filename}")
+
+    ########## UPSERT RECORDS #############
+
+    def merge_upsert_records(self, release: PubMedRelease, **kwargs):
+        """
+        Merge the upsert records that will be applied to the main table.
+
+        Only keep the newest upsert record from the updatefiles.
+        """
+
+        upsert_files_to_merge = [datafile for datafile in release.datafile_list if not datafile.baseline]
+
+        upsert_files = {}
+        for datafile in upsert_files_to_merge:
+            transform_file = datafile.transform_file_path(entity_type=self.entity_list["upsert"].type)
+
+            with open(transform_file, "r") as f_in:
+                incoming_data = [json.loads(line) for line in f_in]
```

Comment:
There are pre-made functions `from observatory.platform.files import load_jsonl, save_jsonl_gz, save_jsonl` that can be used to load and save a lot of the files.

```suggestion
            transform_file = datafile.transform_file_path(entity_type=self.entity_list["upsert"].type)
            incoming_data = load_jsonl(transform_file)
```